core/modules/map.rs

```diff
@@ -436,11 +502,116 @@ impl ModuleMap {
     Ok(id)
   }
 
+  /// Create and compile an ES module.
+  pub(crate) fn new_wasm_module(
+    &self,
+    scope: &mut v8::HandleScope,
+    main: bool,
+    name: ModuleName,
+    source: ModuleCodeString,
+    is_dynamic_import: bool,
+  ) -> Result<ModuleId, ModuleError> {
```

Comment:
This should be abstracted into a separate function and upstreamed in a separate PR. Maybe `new_es_module` should accept a `ModuleType` to define how the module introduces itself.